### PR TITLE
Add a conditional indexer

### PR DESCRIPTION
**Pull request by dadgar**
Add a conditional indexer that lets a user specify a function that returns true or false based on the passed object and create an arbitrarily complex index.

### CODE REVIEW

**Comment by dadgar** [comment]
@slackpad Could you please review.

File: index.go [hunk 1]
```diff
@@ -200,6 +200,42 @@ func (f *FieldSetIndex) FromObject(obj interface{}) (bool, []byte, error) {
...
+type ConditionalIndex struct {
+	Conditional ConditionalIndexFunc
+}
+
+// ConditionalIndexFunc is the required function interface for a
+// ConditionalIndex.
+type ConditionalIndexFunc func(obj interface{}) (bool, error)
+
+func (c *ConditionalIndex) FromObject(obj interface{}) (bool, []byte, error) {
+	// Call the users function
```
Comment (slackpad) [inline index.go]: *user's*

File: index_test.go [hunk 2]
```diff
@@ -316,6 +316,86 @@ func TestFieldSetIndex_FromArgs(t *testing.T) {
...
+	}
+	if !ok {
+		t.Fatalf("should be ok")
+	}
+
+	// Pass an invalid type.
+	ok, val, err = indexer.FromObject(t)
+	if err == nil {
+		t.Fatalf("expected an error when passing invalid type")
+	}
```
Comment (slackpad) [inline index_test.go]: should verify `!ok` just for completeness

File: index_test.go [hunk 3]
```diff
@@ -316,6 +316,86 @@ func TestFieldSetIndex_FromArgs(t *testing.T) {
...
+	indexer := ConditionalIndex{conditional}
+	obj.Bar = 42
+	ok, val, err := indexer.FromObject(obj)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if len(val) != 1 || val[0] != 1 {
+		t.Fatalf("bad: %v", val)
+	}
+	if !ok {
```
Comment (slackpad) [inline index_test.go]: just a style thing, i'd always check OK first

**Comment by slackpad** [comment]
A couple tiny things, otherwise LGTM.

**Pull request merged**